### PR TITLE
provide more helpful error message when `run`ning an improperly-typed term

### LINKED ISSUE #1747: `run` reports that a given term does not exist if that term does not have the proper type
Here's a bad attempt at writing a `run`able function:
```
iExistIPromise = 'printLine "asdf"
```
`ucm` will tell me, both before and after I `add` it to the codebase, that this term does not exist if I try to `run` it:
```
.scratch> load

  I found and typechecked these definitions in ~/rummikub/scratch.u. If you do an `add` or
  `update`, here's how your codebase would change:
  
    ⍟ These new definitions are ok to `add`:
    
      iExistIPromise : Text ->{IO} ()

.scratch> run iExistIPromise

  😶
  
  I looked for a function `iExistIPromise` in the most recently typechecked file and codebase but
  couldn't find one. It has to have the type:
  
    iExistIPromise : '{.base.io.IO} ()

.scratch> add iExistIPromise

  ⍟ I've added these definitions:
  
    iExistIPromise : Text ->{IO} ()


  I added some default metadata.

.scratch> run iExistIPromise

  😶
  
  I looked for a function `iExistIPromise` in the most recently typechecked file and codebase but
  couldn't find one. It has to have the type:
  
    iExistIPromise : '{.base.io.IO} ()
```

This error message leads me to believe it's failing to even detect the existence of that term. I believe what's actually happening is it cannot find that term _with the proper type_, which is `'{.base.io.IO} ()`. If I update the function to have that proper type:
```
iExistIPromise = '(handle printLine "asdf" with IOCLIHandler)
```

Then I can successfully `run` it, even before `add`ing it to the codebase:
```
.scratch> 

  I found and typechecked these definitions in ~/rummikub/scratch.u. If you do an `add` or
  `update`, here's how your codebase would change:
  
    ⍟ These names already exist. You can `update` them to your new definition:
    
      iExistIPromise : ∀ (). () ->{IO} ()

.scratch> run iExistIPromise
asdf
```

### PR DESCRIPTION
**Pull request by samgqroberts**
## Overview

Endeavoring to fix #1747 - see the output of the new transcript for what the new error message is and when it is shown.

## Implementation notes

Instead of `HandleInput.addRunMain` just returning a `Maybe` (I could do it, or I couldn't), it now returns a type that disambiguates whether it couldn't due to not being able to find the term, or whether the term has a bad type. That new case is subsequently handled in the error reporting module `OutputMessages.hs` to produce the new error message.

## Test coverage

I've included a new transcript `run.md` that attempts to `run` properly-typed, improperly-typed, and nonexistent terms to show the differences in behavior / error messages.

## Loose ends

- **This PR is obfuscated by another bug** - you'll see me call this out in the new transcript. I believe I'm running into #1800, and I've confirmed that without my changes that `runnable` function still fails to be found. I wanted to put this PR up optimistically to gather feedback while that issue is resolved.
- I didn't update the error reporting in `Execute.hs` to provide this distinction (partly due to my lack of familiarity of what experience that file is part of..)

In general, this being my first contribution, I'm not confident at all that the changes I've provided are idiomatic / sufficient. I'd be more than happy to add tests, move code, rename types. Thank you!

### CODE REVIEW

**Comment by pchiusano** [comment]
This is great, a nice little usability improvement! Can you add yourself to `CONTRIBUTORS.markdown`? I'll wait for CI to pass before merging.

**Review by aryairani — APPROVED** [review]
LGTM

**Pull request merged**
